### PR TITLE
Add decomposition for MultiControlledX gate

### PR DESCRIPTION
**Pull request by trbromley**
The `MultiControlledX` gate can now be decomposed, allowing for it to potentially be implemented on hardware or other devices that don't support it.

```python
ctrl_wires = [f"c{i}" for i in range(5)]
work_wires = ["w1"]
target_wires = ["t1"]
all_wires = ctrl_wires + work_wires + target_wires

dev = qml.device("default.qubit", wires=all_wires)

with qml.tape.QuantumTape() as tape:
    qml.MultiControlledX(control_wires=ctrl_wires, wires=target_wires, work_wires=work_wires)
```

```pycon
>>> tape = tape.expand(depth=2)
>>> print(tape.draw(wire_order=Wires(all_wires)))
 c0: ──────────╭C──────────────╭C──────────────────╭C──────────────╭C──────────┤  
 c1: ──────────├C──────────────├C──────────────────├C──────────────├C──────────┤  
 c2: ──────╭C──│───╭C──────╭C──│───╭C──────────╭C──│───╭C──────╭C──│───╭C──────┤  
 c3: ──╭C──│───│───│───╭C──│───│───│───────╭C──│───│───│───╭C──│───│───│───────┤  
 c4: ──│───├C──╰X──├C──│───├C──╰X──├C──╭C──│───├C──╰X──├C──│───├C──╰X──├C──╭C──┤  
 w0: ──├X──│───────│───├X──│───────│───├C──├X──│───────│───├X──│───────│───├C──┤  
 t1: ──╰C──╰X──────╰X──╰C──╰X──────╰X──╰X──╰C──╰X──────╰X──╰C──╰X──────╰X──╰X──┤ 
```

### CODE REVIEW

File: pennylane/ops/qubit.py [hunk 1]
```diff
@@ -1975,15 +1976,19 @@ def __init__(self, *params, control_wires=None, wires=None, control_values=None,
...
-        params = list(params)
-        params[0] = CU
+    def _matrix(self, *params):
+        if self._CU is None:
+            self._CU = block_diag(np.eye(self._padding_left), self.U, np.eye(self._padding_right))
 
-        super().__init__(*params, wires=wires, do_queue=do_queue)
+        params = list(params)
+        params[0] = self._CU
+        return super()._matrix(*params)
```
Comment (trbromley) [inline pennylane/ops/qubit.py:1991]: Here we are changing `ControlledQubitUnitary` so that the full controlled matrix is not loaded upon initialization but instead when the `matrix` attribute is used. Caching is used, so that the matrix only need be constructed once.

The motivation for doing this is that `MultiControlledX` (which inherits from `ControlledQubitUnitary`) could be limited in size due to memory requirements because (without this change) `ControlledQubitUnitary` immediately loads the matrix, even if it's not needed because the op will be decomposed.

In other words, with this change we're able to create much bigger `MultiControlledX` gates and defer the simulation to the device (which will either use the matrix or the decomp).

File: pennylane/ops/qubit.py [hunk 2]
```diff
@@ -2051,6 +2080,107 @@ def __init__(self, control_wires=None, wires=None, control_values=None, do_queue
             do_queue=do_queue,
         )
 
+    # pylint: disable=unused-argument
+    def decomposition(self, *args, **kwargs):
```
Comment (trbromley) [inline pennylane/ops/qubit.py:2084]: Note that `decomposition` is not a static method. We have access to `self` and also don't care about `*args, **kwargs` because we read the relevant data from the object's state (e.g., `self.control_wires`).

This is going against the standard, but it seems to work.

File: pennylane/ops/qubit.py [hunk 3]
```diff
@@ -2051,6 +2080,107 @@ def __init__(self, control_wires=None, wires=None, control_values=None, do_queue
...
+
+        if len(self.control_wires) > 2 and len(self._work_wires) == 0:
+            raise ValueError(f"At least one work wire is required to decompose operation: {self}")
```
Comment (trbromley) [inline pennylane/ops/qubit.py:2087]: We could add other decompositions in future if we find alternatives.

File: pennylane/ops/qubit.py [hunk 4]
```diff
@@ -2021,6 +2026,15 @@ class MultiControlledX(ControlledQubitUnitary):
...
         control_values (str): a string of bits representing the state of the control
             qubits to control on (default is the all 1s state)
+        work_wires (Union[Wires, Sequence[int], or int]): optional work wires used to decompose
+            the operation into a series of Toffoli gates
+
+    .. note::
+
+        If ``MultiControlledX`` is not supported on the targeted device, PennyLane will decompose
+        ``MultiControlledX`` into Toffoli gates using the methods described in Lemma 7.2 and 7.3 of
+        `Barenco et al. <https://arxiv.org/abs/quant-ph/9503016>`__. For control on 3 or more wires,
```
Comment (trbromley) [inline pennylane/ops/qubit.py:2036]:
```suggestion
        `Barenco et al. <https://arxiv.org/abs/quant-ph/9503016>`__ For control on 3 or more wires,
```

File: tests/ops/test_qubit_ops.py [hunk 5]
```diff
@@ -2097,6 +2099,190 @@ def circuit_pauli_x():
...
+            qml.BasisState(bitstring, wires=range(n_ctrl_wires + 1))
+            qml.MultiControlledX(
+                control_wires=control_wires, wires=target_wire, control_values=control_values
+            ).inv()
+            for op in tape.operations:
+                op.queue()
+            return qml.probs(wires=range(n_ctrl_wires + 1))
+
+        u = np.array([f(b) for b in itertools.product(range(2), repeat=n_ctrl_wires + 1)]).T
+        spy.assert_called()
```
Comment (DSGuala) [inline tests/ops/test_qubit_ops.py:2219]: Very interesting. I didn't know about `spy.assert_called()`

**Review by DSGuala — APPROVED** [review]

File: pennylane/ops/qubit.py [hunk 6]
```diff
@@ -2051,6 +2082,113 @@ def __init__(self, control_wires=None, wires=None, control_values=None, do_queue
...
+        flips2 = [
+            qml.PauliX(self.control_wires[i])
+            for i, val in enumerate(self.control_values)
+            if val == "0"
+        ]
+
+        return flips1 + decomp + flips2
+
+    @staticmethod
+    def _decomposition_with_many_workers(control_wires, target_wire, work_wires):
```
[the remark this answers is not in the document]
Comment (DSGuala) [reply pennylane/ops/qubit.py:2123]: My thoughts are that since `w0`,`w1`, and `w2` are work qubits that start in `|0>` these CCX gates in the beginning will not have an effect on `t1` or the work qubits. Then we wouldn't need to uncompute them at the end either.

File: pennylane/ops/qubit.py [hunk 7]
```diff
@@ -2051,6 +2082,113 @@ def __init__(self, control_wires=None, wires=None, control_values=None, do_queue
...
+        """Decomposes the multi-controlled PauliX gate using the approach in Lemma 7.3 of
+        https://arxiv.org/pdf/quant-ph/9503016.pdf, which requires a single work wire"""
+        tot_wires = len(control_wires) + 2
+        partition = int(np.ceil(tot_wires / 2))
+
+        first_part = control_wires[:partition]
+        second_part = control_wires[partition:]
+
+        gates = [
+            MultiControlledX(
```
[the remark this answers is not in the document]
Comment (glassnotes) [reply pennylane/ops/qubit.py:2176]: They do need to be decomposed, but the n-controlled operation is broken down into 4 smaller m-controlled operations such that there are always enough unused qubits that they can be used as work qubits. Like in the example:
image
With the extra work qubit there, the 7-controlled gate turns into a 5-controlled gate and there are conveniently 3 leftover qubits that could be used as scratch space.

That said, yes, there is a tradeoff because you can tinker with the relative sizes of the four smaller `MultiControlledX` gates. I think the optimum is ceil(n / 2) (which is what gets used here).

File: tests/ops/test_qubit_ops.py [hunk 8]
```diff
@@ -2097,6 +2099,190 @@ def circuit_pauli_x():
...
+        assert np.allclose(u, np.eye(2 ** (n_ctrl_wires + 1)))
+
+    def test_worker_state_unperturbed(self, mocker):
+        """Test that the state of the worker wires is unperturbed after the decomposition has used
+        them. To do this, a random state over all the qubits (control, target and workers) is
+        loaded and U^dagger U(decomposed) is applied. If the workers are uncomputed, the output
+        state will be the same as the input."""
```
Comment (glassnotes) [inline tests/ops/test_qubit_ops.py:2258]: This is a great test to have!

**Review by glassnotes — COMMENTED** [review]
This looks great! Just left some small questions and comments.

File: .github/CHANGELOG.md [hunk 9]
```diff
@@ -108,6 +108,34 @@ random_mat2 = rng.standard_normal(3, requires_grad=False)
...
+  dev = qml.device("default.qubit", wires=all_wires)
+
+  with qml.tape.QuantumTape() as tape:
+      qml.MultiControlledX(control_wires=ctrl_wires, wires=target_wires, work_wires=work_wires)
+  ```
+  
+  ```pycon
+  >>> tape = tape.expand(depth=2)
+  >>> print(tape.draw(wire_order=Wires(all_wires)))
+   c0: ──────────╭C──────────────╭C──────────────────╭C──────────────╭C──────────┤
```
[the remark this answers is not in the document]
Comment (trbromley) [reply .github/CHANGELOG.md]: image
^ For reference

File: pennylane/ops/qubit.py [hunk 10]
```diff
@@ -2021,6 +2028,15 @@ class MultiControlledX(ControlledQubitUnitary):
...
             qubits to control on (default is the all 1s state)
+        work_wires (Union[Wires, Sequence[int], or int]): optional work wires used to decompose
+            the operation into a series of Toffoli gates
+
+    .. note::
+
+        If ``MultiControlledX`` is not supported on the targeted device, PennyLane will decompose
+        ``MultiControlledX`` into Toffoli gates using the methods described in Lemma 7.2 and 7.3 of
+        `Barenco et al. <https://arxiv.org/abs/quant-ph/9503016>`__ For control on 3 or more wires,
+        at least one work wire will be required.
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/ops/qubit.py]: Thanks, have done here: 863fc045e65550223784b4f6497e9d0394b4acbe

File: pennylane/ops/qubit.py [hunk 11]
```diff
@@ -2051,6 +2082,113 @@ def __init__(self, control_wires=None, wires=None, control_values=None, do_queue
...
+
+    @staticmethod
+    def _decomposition_with_many_workers(control_wires, target_wire, work_wires):
+        """Decomposes the multi-controlled PauliX gate using the approach in Lemma 7.2 of
+        https://arxiv.org/pdf/quant-ph/9503016.pdf, which requires a suitably large register of
+        work wires"""
+        num_work_wires_needed = len(control_wires) - 2
+        work_wires = work_wires[:num_work_wires_needed]
+
+        work_wires_reversed = list(reversed(work_wires))
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/ops/qubit.py:2144]: Edit: started trying this but it made my brain hurt :laughing:

File: pennylane/circuit_drawer/representation_resolver.py [hunk 12]
```diff
@@ -376,6 +376,11 @@ def operator_representation(self, op, wire):
                 op, "U", self.unitary_matrix_cache
             )
 
+        elif base_name == "MultiControlledX":
+            if wire in op.control_wires:
+                return self.charset.CONTROL
+            representation = "X"
+
```
[the remark this answers is not in the document]
Comment (trbromley) [reply pennylane/circuit_drawer/representation_resolver.py:383]: One caveat is that I couldn't think of a way to represent `control_values` :thinking: But I think this can be left as a follow-up question.

File: pennylane/ops/qubit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DSGuala) [reply pennylane/ops/qubit.py:2139]: I hadn't realized that the work wires can be in a non-|0> state and still give the right results! That's exciting resource-wise!

**Review by ixfoduap — APPROVED** [review]
:shipit:
